File: dataStructure/binarySearchTree/code.py
class Node(object):
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BinarySearchTree(object):
    def __init__(self):
        self.head = None

    def getHead(self):
        return self.head

    def insert(self, node):
        if self.head == None:
            self.head = node
        else:
            current = self.head
            while True:
                if current.value > node.value:
                    if current.left == None:
                        current.left = node
                        break
                    else:
                        current = current.left
                else:
                    if current.right == None:
                        current.right = node
                        break
                    else:
                        current = current.right

    def search(self, value):
        current = self.head
        while current:
            if current.value == value:
                return True, current
            elif current.value > value:
                current = current.left
            else:
                current = current.right
        return False, None

    # def __removeValue():

    def remove(self, value):
        searched = False
        parent = self.head
        current = self.head
        while current:
            if current.value == value:
                searched = True
                break
            elif current.value > value:
                parent = current
                current = current.left
            else:
                parent = current
                current = current.right

        if searched == False:
            return False
        else:
            if current == self.head:
                d
            elif current.left == None and current.right == None:
                if parent.value > value:
                    parent.left = None
                else:
                    parent.right = None
            elif current.right == None:
                if parent.value > value:
                    parent.left = current.left
                else:
                    parent.right = current.left
            elif current.left == None:
                if parent.value > value:
                    parent.left = current.right
                else:
                    parent.right = current.right
            else:
                if parent.value > value:
                    changeParent = current.right
                    change = current.right
                    while change.left:
                        changeParent = change
                        change = change.left
                    if change.right != None:
                        changeParent.left = change.right
                    else:
                        changeParent.left = None
                    parent.left = change
                    change.right = current.right
                    change.left = current.left
                else:
                    changeParent = current.right
                    change = current.right
                    while change.left:
                        changeParent = change
                        change = change.left
                    if change.right != None:
                        changeParent.left = change.right
                    else:
                        changeParent.left = None
                    parent.right = change
                    change.right = current.right
                    change.left = current.left


bst = BinarySearchTree()
bst.insert(Node(10))
bst.insert(Node(15))
bst.insert(Node(5))
bst.insert(Node(20))
bst.insert(Node(13))
bst.insert(Node(1))
bst.insert(Node(6))
bool, node = bst.search(10)
if node:
    print(node.value, ": left ->", node.left.value if node.left != None else None)
    print(node.value, ": right ->", node.right.value if node.right != None else None)
bool, node = bst.search(15)
if node:
    print(node.value, ": left ->", node.left.value if node.left != None else None)
    print(node.value, ": right ->", node.right.value if node.right != None else None)
bool, node = bst.search(5)
if node:
    print(node.value, ": left ->", node.left.value if node.left != None else None)
    print(node.value, ": right ->", node.right.value if node.right != None else None)

File: dataStructure/binarySearchTree/code.js
class Node {
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left = left;
    this.right = right;
  }
}

class BinarySearchThree {
  #head = null;

  getHead() {
    return this.#head;
  }

  insert(node) {
    if (!this.#head) {
      this.#head = node;
      return;
    }
    let current = this.#head;
    while (true) {
      if (node.value < current.value) {
        if (!current.left) {
          current.left = node;
          break;
        } else current = current.left;
      } else {
        if (!current.right) {
          current.right = node;
          break;
        } else current = current.right;
      }
    }
  }

  searchAndGetNode(value) {
    let current = this.#head;
    while (current) {
      if (current.value == value) return current;
      else if (value < current.value) current = current.left;
      else current = current.right;
    }
    return null;
  }

  #removeValue(current, parent, change, chParent) {
    while (change.left) {
      chParent = change;
      change = change.left;
    }
    if (current.right === change) {
      parent.right = change;
      change.left = current.left;
      return true;
    }
    if (!change.right) chParent.left = change.right;
    else chParent.left = null;
    return false;
  }

  remove(value) {
    let current = this.#head;
    let parent = new Node(Infinity, this.#head, this.#head);
    let searched = false;
    while (current) {
      if (value == current.value) {
        searched = true;
        break;
      } else if (value < current.value) {
        parent = current;
        current = current.left;
      } else {
        parent = current;
        current = current.right;
      }
    }
    if (searched) {
      if (!current.left && !current.right) {
        if (value < parent.value) parent.left = null;
        else parent.right = null;
      } else if (!current.left) {
        if (value < parent.value) parent.left = current.right;
        else parent.right = current.right;
      } else if (!current.right) {
        if (value < parent.value) parent.left = current.left;
        else parent.right = current.left;
      } else {
        let change = current.right;
        let chParent = current.right;
        if (value < parent.value) {
          const flag = this.#removeValue(current, parent, change, chParent);
          if (!flag) {
            parent.left = change;
            change.right = current.right;
            change.left = current.left;
          }
        } else {
          const flag = this.#removeValue(current, parent, change, chParent);
          if (!flag) {
            parent.right = change;
            change.right = current.right;
            change.left = current.left;
          }
        }
      }
    }
  }
}

const bst = new BinarySearchThree();
bst.insert(new Node(10));
bst.insert(new Node(5));
bst.insert(new Node(15));
bst.insert(new Node(17));
bst.insert(new Node(13));
bst.insert(new Node(7));
bst.insert(new Node(3));
console.log(bst.getHead());
bst.remove(15);
console.log(bst.getHead());
bst.remove(10);
console.log(bst.searchAndGetNode(10));
console.log(bst.getHead());
